sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -188,6 +189,12 @@ void stop() {
             attemptsLeft);
       }
     }
+    if (!isShutdown) {
+      LOG.warn(
+          "Executor service was not properly shutdown after {} attempts, trying to forcibly stop it.",
+          ATTEMPTS_TO_SHUTDOWN);
+      executorService.shutdownNow();
```

Comment:
I think that we need to repeat the `awaitTermination` loop after forcible `shutdownNow`. It's possible that invoking `shutdown` won't have any effect within 30s timeout as it's not interrupting currently running tasks (that's why it wasn't originally used). So this approach won't interrupt the `readLoop` when it's waiting for Kinesis to return some records or when putting records to a full recordsQueue. But with additional waiting it should be fine.

Reply:
Thanks for this note, agree with that. So, I added another `awaitTermination` loop and did a bit of refactoring of this part to make code easier to read.

Reply:
Great, it LGTM.